### PR TITLE
[MAINT] ensure warning is thrown if estimators get non finite values in data to transform

### PR DESCRIPTION
**Pull request by Remi-Gau**
- Closes none but relates #4538

Changes proposed in this pull request:

- adds an equivalent to sklearn check_estimators_nan_inf

https://github.com/scikit-learn/scikit-learn/blob/19a6e61b6f8a7bb8a6d9dd5e8a5d40a741de28c3/sklearn/utils/estimator_checks.py#L2315

### CODE REVIEW

File: nilearn/image/resampling.py [hunk 1]
```diff
@@ -284,7 +284,7 @@ def _resample_one_img(
             # We need to do a copy to avoid modifying the input
             # array
             data = data.copy()
-        # data[not_finite] = 0
```
Comment (Remi-Gau) [inline nilearn/image/resampling.py:287]: remove commented code

File: nilearn/_utils/ndimage.py [hunk 2]
```diff
@@ -160,3 +164,18 @@ def peak_local_max(
     nd_indices = tuple(coordinates.T)
     out[nd_indices] = True
     return out
+
+
+def replace_non_finite(data, value=None):
```
Comment (Remi-Gau) [inline nilearn/_utils/ndimage.py:169]: introduce a single function to handle non finite data in several places

**Comment by bthirion** [comment]
It's not yet completely clear to me how this relates to #4538 
Do you mean that the warn+impute is the current or recommended strategy in sklearn ?

**Comment by Remi-Gau** [comment]
> It's not yet completely clear to me how this relates to #4538 Do you mean that the warn+impute is the current or recommended strategy in sklearn ?

The issues keeps track of the sklearn estimator checks and those that are applied or have a replacement in nilearn.

For this PR, several sklearn estimators throw errors if the data to fit / predict / transform contains non finite values.

This is the sklearn estimator checks that make sure this behavior is enforced:
https://github.com/scikit-learn/scikit-learn/blob/19a6e61b6f8a7bb8a6d9dd5e8a5d40a741de28c3/sklearn/utils/estimator_checks.py#L2315

Throwing errors feels too stringent for nilearn, but at least throwing some warnings would be good.

So the PR just introduces an alternative to the sklearn check and only ensure that warnings are thrown.

**Comment by Remi-Gau** [comment]
- blocked by #5403 

will be better to also implement this on decoders so let's wait for  #5403  to be done

**Review by bthirion — COMMENTED** [review]
The changeset LGTM but I think that this behavior needs to be documented very explicitly

**Comment by Remi-Gau** [comment]
> The changeset LGTM but I think that this behavior needs to be documented very explicitly

Actually I realize that the behavior may not be consistent across estimators.
Some maskers replace in the non finite value with zeros at transform time, while others only throw a warning.

I need to double check if this inconsistency is recent or not.

My hunch is that we probably should not decide for the user what to replace the values with, so only throw warning sounds more cautious.

**Comment by Remi-Gau** [comment]
little script to check the behavior of the nifti maskers when dealing with images with
- finite data
- data with nans
- data with inf

```python
import warnings
from pathlib import Path

import numpy as np
import pandas as pd
from nibabel import Nifti1Image
from sklearn import clone

from nilearn import __version__
from nilearn._utils.data_gen import generate_labeled_regions, generate_maps
from nilearn.maskers import NiftiLabelsMasker, NiftiMapsMasker, NiftiMasker

shape = (4, 4, 4)
affine = np.eye(4)

data = np.ones(shape, dtype="float32")
assert np.all(np.isfinite(data))
img_finite = Nifti1Image(data, affine, dtype="float32")

data[1, 2, 2] = np.nan
data[2, 3, 1] = np.nan
assert not np.all(np.isfinite(data))
img_nan = Nifti1Image(data, affine, dtype="float32")

data[1, 2, 2] = np.inf
data[2, 3, 1] = np.inf
assert not np.all(np.isfinite(data))
img_inf = Nifti1Image(data, affine, dtype="float32")

mask_img = Nifti1Image(np.ones(shape, dtype="float32"), affine)

labels_img = generate_labeled_regions(shape, n_regions=2, affine=affine)

maps_img, _ = generate_maps(shape, n_regions=2, affine=affine)

maskers = [
    NiftiMasker(mask_img),
    NiftiMasker(),
    NiftiLabelsMasker(labels_img=labels_img, mask_img=mask_img),
    NiftiLabelsMasker(labels_img=labels_img),
    NiftiMapsMasker(maps_img=maps_img, mask_img=mask_img),
    NiftiMapsMasker(maps_img=maps_img),
]

data = {
    "masker": [],
    "has_mask_img": [],
    "img_content": [],
    "signal_shape": [],
    "singal_size": [],
    "fit_warning": [],
    "transform_warning": [],
}

for masker in maskers:
    for img, content in zip([img_finite, img_nan, img_inf], ["finite", "nan", "inf"]):
        mask_str = "without" if masker.mask_img is None else "with"
        print(f"\n{masker.__class__.__name__} {mask_str} mask - {content}")

        masker = clone(masker)

        with warnings.catch_warnings(record=True) as w_f:
            masker.fit(img)

        with warnings.catch_warnings(record=True) as w_t:
            signal = masker.transform(img)

        print(f"{signal.shape=}")
        print(f"{signal.size=}")
        assert np.all(np.isfinite(signal))

        data["masker"].append(masker.__class__.__name__)
        data["has_mask_img"].append(masker.mask_img is None)
        data["img_content"].append(content)
        data["signal_shape"].append(signal.shape)
        data["singal_size"].append(signal.size)
        data["fit_warning"].append(len(w_f))
        data["transform_warning"].append(len(w_t))

df = pd.DataFrame(data)
print(df.to_markdown())
```

These are the results from 0.9.2 to 0.11.1:

not sure why NiftiMasker output of transform is affected this much by non finite data

so NiftiLabelsMasker warns about something being non-finite when passing a finite data with a finite mask, but not in any other situation... Great...

| masker            | has_mask_img   | content   | shape   |   size |   fit_warning |   trans_warning |
|:------------------|:---------------|:----------|:--------|-------:|--------------:|----------------:|
| NiftiMasker       | True           | finite    | (1, 64) |     64 |             0 |               0 |
| NiftiMasker       | True           | nan       | (1, 64) |     64 |             0 |               0 |
| NiftiMasker       | True           | inf       | (1, 64) |     64 |             0 |               0 |
| NiftiMasker       | False          | finite    | (1, 2)  |      2 |             0 |               0 |
| NiftiMasker       | False          | nan       | (1, 2)  |      2 |             0 |               0 |
| NiftiMasker       | False          | inf       | (1, 2)  |      2 |             0 |               0 |
| NiftiLabelsMasker | True           | finite    | (1, 2)  |      2 |             0 |               1 |
| NiftiLabelsMasker | True           | nan       | (1, 2)  |      2 |             0 |               0 |
| NiftiLabelsMasker | True           | inf       | (1, 2)  |      2 |             0 |               0 |
| NiftiLabelsMasker | False          | finite    | (1, 2)  |      2 |             0 |               0 |
| NiftiLabelsMasker | False          | nan       | (1, 2)  |      2 |             0 |               0 |
| NiftiLabelsMasker | False          | inf       | (1, 2)  |      2 |             0 |               0 |
| NiftiMapsMasker   | True           | finite    | (1, 2)  |      2 |             0 |               0 |
| NiftiMapsMasker   | True           | nan       | (1, 2)  |      2 |             0 |               0 |
| NiftiMapsMasker   | True           | inf       | (1, 2)  |      2 |             0 |               0 |
| NiftiMapsMasker   | False          | finite    | (1, 2)  |      2 |             0 |               0 |
| NiftiMapsMasker   | False          | nan       | (1, 2)  |      2 |             0 |               0 |
| NiftiMapsMasker   | False          | inf       | (1, 2)  |      2 |             0 |               0 |

**Comment by Remi-Gau** [comment]
That's on main:

the "new" warning for NiftiMasker initialised with a mask_img are not related to non finite data but to the fact that the masker now appropriately tell users that when the mask_img passed at init will have 'priority'

the change of shape of the singal returned by transform is to be expected because we are transforming a 3D image (this new behavior was planned for 2 versions)

| masker            | has_mask_img   | content   | shape   |   size |   fit_warning |   trans_warning |
|:------------------|:---------------|:----------|:--------|-------:|--------------:|----------------:|
| NiftiMasker       | True           | finite    | (64,)   |     64 |             1 |               0 |
| NiftiMasker       | True           | nan       | (64,)   |     64 |             1 |               0 |
| NiftiMasker       | True           | inf       | (64,)   |     64 |             1 |               0 |
| NiftiMasker       | False          | finite    | (2,)    |      2 |             0 |               0 |
| NiftiMasker       | False          | nan       | (2,)    |      2 |             0 |               0 |
| NiftiMasker       | False          | inf       | (2,)    |      2 |             0 |               0 |
| NiftiLabelsMasker | True           | finite    | (2,)    |      2 |             0 |               1 |
| NiftiLabelsMasker | True           | nan       | (2,)    |      2 |             0 |               0 |
| NiftiLabelsMasker | True           | inf       | (2,)    |      2 |             0 |               0 |
| NiftiLabelsMasker | False          | finite    | (2,)    |      2 |             0 |               0 |
| NiftiLabelsMasker | False          | nan       | (2,)    |      2 |             0 |               0 |
| NiftiLabelsMasker | False          | inf       | (2,)    |      2 |             0 |               0 |
| NiftiMapsMasker   | True           | finite    | (2,)    |      2 |             0 |               0 |
| NiftiMapsMasker   | True           | nan       | (2,)    |      2 |             0 |               0 |
| NiftiMapsMasker   | True           | inf       | (2,)    |      2 |             0 |               0 |
| NiftiMapsMasker   | False          | finite    | (2,)    |      2 |             0 |               0 |
| NiftiMapsMasker   | False          | nan       | (2,)    |      2 |             0 |               0 |
| NiftiMapsMasker   | False          | inf       | (2,)    |      2 |             0 |               0 |

**Comment by Remi-Gau** [comment]
and that's on this PR for now:

| masker            | has_mask_img   | content   | shape   |   size |   fit_warning |   trans_warning |
|:------------------|:---------------|:----------|:--------|-------:|--------------:|----------------:|
| NiftiMasker       | True           | finite    | (64,)   |     64 |             1 |               2 |
| NiftiMasker       | True           | nan       | (64,)   |     64 |             1 |               2 |
| NiftiMasker       | True           | inf       | (64,)   |     64 |             1 |               2 |
| NiftiMasker       | False          | finite    | (2,)    |      2 |             0 |               2 |
| NiftiMasker       | False          | nan       | (2,)    |      2 |             0 |               2 |
| NiftiMasker       | False          | inf       | (2,)    |      2 |             0 |               2 |
| NiftiLabelsMasker | True           | finite    | (2,)    |      2 |             0 |               1 |
| NiftiLabelsMasker | True           | nan       | (2,)    |      2 |             0 |               0 |
| NiftiLabelsMasker | True           | inf       | (2,)    |      2 |             0 |               0 |
| NiftiLabelsMasker | False          | finite    | (2,)    |      2 |             0 |               0 |
| NiftiLabelsMasker | False          | nan       | (2,)    |      2 |             0 |               0 |
| NiftiLabelsMasker | False          | inf       | (2,)    |      2 |             0 |               0 |
| NiftiMapsMasker   | True           | finite    | (2,)    |      2 |             0 |               0 |
| NiftiMapsMasker   | True           | nan       | (2,)    |      2 |             0 |               0 |
| NiftiMapsMasker   | True           | inf       | (2,)    |      2 |             0 |               0 |
| NiftiMapsMasker   | False          | finite    | (2,)    |      2 |             0 |               0 |
| NiftiMapsMasker   | False          | nan       | (2,)    |      2 |             0 |               0 |
| NiftiMapsMasker   | False          | inf       | (2,)    |      2 |             0 |               0 |

**Comment by Remi-Gau** [comment]
conclusion for now:
- this PR does not seem to do what it's supposed to do
- not sure why niftimaskers transform output is affected by non finite data

**Comment by Remi-Gau** [comment]
removed this one from the 0.12 milestone: may take some time to figure out what we want the behavior to be here

**Comment by bthirion** [comment]
Is there some deep issue behind this ? I thought you had mostly done it.

**Comment by Remi-Gau** [comment]
> Is there some deep issue behind this ? I thought you had mostly done it.

- I thought so too, but it NiftiMasker seems to be doing something 'funky' with non finite data (even before this PR) and I want to understand it better before I feel comfortable about merging this
- also the little script I wrote to compare behavior of maskers before this PR and after seems to show that the PR is failing at what it is supposed to do, so anther thing I need to figure out

don't have time to check now, so I will get back to it after 0.12

**Comment by bthirion** [comment]
LMK when you want some more feedback on this (indeed the maskers' behavior seems tricky).

**Comment by Remi-Gau** [comment]
not had time to circle back to this one: also I am avoiding it like the monster under my bed but I will get back to it

**Comment by Remi-Gau** [comment]
Need to double check if this was not closed prematurely.
